bleach/sanitizer.py

```diff
@@ -588,3 +591,57 @@ def sanitize_css(self, style):
                 clean.append(prop + ': ' + value + ';')
 
         return ' '.join(clean)
+
+
+class BleachHTMLSerializer(HTMLSerializer):
+    """Wraps the HTMLSerializer and undoes & -> &amp; in attributes"""
+    def escape_base_amp(self, stoken):
+        """Escapes bare & in HTML attribute values"""
+        # First, undo what the HTMLSerializer did
+        stoken = stoken.replace('&amp;', '&')
+
+        # Then, escape any bare &
+        for part in next_possible_entity(stoken):
+            if not part:
+                continue
+
+            if part.startswith('&'):
+                entity = match_entity(part)
+                if entity is not None:
+                    yield '&' + entity + ';'
+
+                    # Length of the entity plus 2--one for & at the beginning
+                    # and and one for ; at the end
+                    part = part[len(entity) + 2:]
+                    if part:
+                        yield part
+                    continue
+
+            yield part.replace('&', '&amp;')
+
+    def serialize(self, treewalker, encoding=None):
+        """Wrap HTMLSerializer.serialize and escape bare & in attributes"""
+        in_tag = False
+        after_equals = False
+
+        for stoken in super(BleachHTMLSerializer, self).serialize(treewalker, encoding):
```

Comment:
`serialize` nicely returns strings one bit at a time. So:

```
<img src="?foo=bar">
```

Gets returned as the following parts which I'm calling "string tokens" or "stoken" (one per line):

```
<img
src
=
"
?foo=bar
"
>
```

We've got quoting set to "always", so an attribute value is always after a `=` followed by a `"`.